swaybar/bar.c

```diff
@@ -270,9 +271,44 @@ static void handle_global(void *data, struct wl_registry *registry,
 	}
 }
 
+static void swaybar_output_free(struct swaybar_output *output) {
+	if (!output) {
+		return;
+	}
+	wlr_log(L_DEBUG, "Removing output %s", output->name);
+	zwlr_layer_surface_v1_destroy(output->layer_surface);
+	wl_surface_destroy(output->surface);
+	wl_output_destroy(output->output);
+	destroy_buffer(&output->buffers[0]);
+	destroy_buffer(&output->buffers[1]);
+	struct swaybar_workspace *ws, *ws_tmp;
+	wl_list_for_each_safe(ws, ws_tmp, &output->workspaces, link) {
+		wl_list_remove(&ws->link);
+		free(ws->name);
+		free(ws);
+	}
+	struct swaybar_hotspot *hotspot, *hotspot_tmp;
+	wl_list_for_each_safe(hotspot, hotspot_tmp, &output->hotspots, link) {
+		if (hotspot->destroy) {
+			hotspot->destroy(hotspot->data);
+		}
+		free(hotspot);
+	}
+	wl_list_remove(&output->link);
+	free(output->name);
+	free(output);
+}
+
 static void handle_global_remove(void *data, struct wl_registry *registry,
 		uint32_t name) {
-	// who cares
+	struct swaybar *bar = data;
+	struct swaybar_output *output, *tmp;
+	wl_list_for_each_safe(output, tmp, &bar->outputs, link) {
+		if (output->wl_name == name) {
+			swaybar_output_free(output);
```

Comment:
It's a good idea to handle output removal, but there's also another way swaybar can get destroyed: with the layer-surface `closed` event. We should handle this one too (since the compositor can close a swaybar without removing an output). There's a `TODO: Deal with hotplugging` about this above.

Reply:
addressed in latest patch